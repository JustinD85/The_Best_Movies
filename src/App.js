import React, { Component } from 'react';
import './App.css';
import {Route, NavLink} from 'react-router-dom';
const Card = <div>Yeup</div>

class App extends Component {
  render() {
    return (
      <div className="App">
        <nav>
          <NavLink to='/recent' component={Card}>Recent Movies</NavLink>
          <NavLink to='/favorites' component={Card}>Favorite Movies</NavLink>
          <NavLink to='/popular' component={Card}>Popular Movies</NavLink>
        </nav>
      </div>
    );
  }
}

export default App;
