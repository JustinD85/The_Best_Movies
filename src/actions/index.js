export const setFilter = (filter) => ({
    type: 'SET_FILTER',
    filter,
})

export const toggleFavorite = (id)=>({
    type:'TOGGLE_FAVORITE',
    id,
})
