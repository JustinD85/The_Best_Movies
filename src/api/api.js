// const url = 'http://localhost:3000/api/'
const url = 'https://falsemotive.io/api/'

export const postData = async (options = '', user) => {
  const response = await fetch(`${url}${options}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(user)
  })

  if (response.ok) {
    return response;
  } else {
    throw new Error(`Error posting data: ${response.status}`);
  }
}

export const fetchData = async (options) => {
  const response = await fetch(`${url}${options}`)
  if (response.ok) {
    return await response.json()
  } else {
    throw new Error('Error fetching data')
  }
}