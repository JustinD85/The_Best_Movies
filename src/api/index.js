import apiKey from './movie_api_key'

// const dbUrl = 'http://localhost:3000/api/'
const dbUrl = 'https://falsemotive.io/api/'
const moviesUrl = 'https://api.themoviedb.org/3/movie/'
export const options = (method, data) => ({
  method,
  headers: {
    'Content-Type': 'application/json',
  },
  body: JSON.stringify(data)
})

export const postData = async (path = '', data) => {
  const response = await fetch(`${dbUrl}${path}`, options('POST', data))
  if (!response.ok) throw new Error(`Could not post`)
  const unfilteredData = await response.json()
  return unfilteredData.data
}

export const deleteData = async (path = '', data) => {
  const response = await fetch(`${dbUrl}${path}`, options('DELETE', data))
  if (!response.ok) throw new Error(`Could not delete`)
  const unfilteredData = await response.json()
  return unfilteredData.results
}

export const getData = async (path) => {
  const response = await fetch(`${dbUrl}${path}`)
  if (!response.ok) throw new Error(`${response.statusText}`)
  const unfilteredData = await response.json()
  return unfilteredData.data
}

export const getMovies = async (path) => {
  const response = await fetch(`${moviesUrl}${path}${apiKey}`)
  if (!response.ok) throw new Error(`${response.statusText}`)
  const unfilteredMovies = await response.json()
  return unfilteredMovies.results
}
