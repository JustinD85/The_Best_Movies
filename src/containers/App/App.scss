.App {
  text-align: center;
  background: #0B161B;
  height:100vh;
}
*{
  margin:0;
  padding:0;
}
