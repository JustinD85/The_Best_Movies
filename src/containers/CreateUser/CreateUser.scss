.CreateUser {
  display: flex;
  height: 50%;
  width: 72%;
  z-index: 1;
  position: absolute;
  left: 22%;
  top: 10%;
  border-radius: 15px;
  background: rgba(0, 0, 0, 0.9);
  color: #ECEBEE;
  padding: 10px 30px 20px 40px;

  .create-your-account {
    color: #0CD0FC;
  }
  
  form {
    margin: auto;
    display: flex;
    flex-direction: column;
    width: 80%;
  }

  .user-input {
    font-size: 14px;
    border-radius: 5px;
    padding: 6px;
    margin: 10px;
  }

  .submit-user {
    width: 30%;
    padding: 5px;
    border-radius: 10px;
    margin: 10px auto;
  }

  a {
    color: #0CD0FC;
  }
}