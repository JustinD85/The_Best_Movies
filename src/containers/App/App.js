import React, { Component } from 'react'
import { Route, withRouter } from 'react-router-dom'
import { connect } from 'react-redux'
import { getMovies, getData } from '../../api/'
import { setMovies, setLoginPrompt, setPopup, setUser, setFavorites } from '../../actions'
import { View, CreateAccountPrompt } from '../../components/'
import { Login, CreateUser } from '../../containers/'
import PropTypes from 'prop-types'
import '../../styles/main.scss'

export class App extends Component {

  setUserFavorites = async (userID) => {
    const retrieveFavPath = `users/${userID}/favorites`
    const favorites = await getData(retrieveFavPath)
    this.props.setFavorites(favorites, userID)
    this.props.setPopup(false)
  }
  
  componentDidMount = async () => {
    if (localStorage.getItem('users')) {
      const { name, id } = JSON.parse(localStorage.getItem('users'))
      this.props.setUser({ name, id })
      this.setUserFavorites(id)
    }
    let movies
    if (!localStorage.getItem('movies')) {
      try {
        movies = await getMovies('now_playing')
        for (let movie of movies) {
          const path = movie.id + '/videos'
          const trailerKeys = await getMovies(path)
          movie.trailer = trailerKeys[0].key
        }
        localStorage.setItem('movies', JSON.stringify(movies))
      } catch (error) {
        console.log('Error getting data')
      }
    }
    else {
      movies = JSON.parse(localStorage.getItem('movies'))
    }
    this.props.setMovies(movies)
  }

  render() {
    const { canPopup, setPopup, setLoginPrompt } = this.props
    return (
      <div className="App">
        <Route path='/login' component={Login} />
        <Route path='/create-user' render={() => <CreateUser setPopup={setPopup} />} />
        <Route path='/' render={() => canPopup && <CreateAccountPrompt
          setPopup={setPopup}
          setLoginPrompt={setLoginPrompt} />} />
        <Route path='/' render={() => <View movies={this.props.movies} />} />
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  movies: state.movies,
  canPopup: state.shouldPromptLogin,
  user: state.user
})

export const mapDispatchToProps = (dispatch) => ({
  setMovies: (movies) => dispatch(setMovies(movies)),
  setUser: (user) => dispatch(setUser(user)),
  setLoginPrompt: (bool) => dispatch(setLoginPrompt(bool)),
  setPopup: (bool) => dispatch(setPopup(bool)),
  setFavorites: (favorites, user_id) => dispatch(setFavorites(favorites, user_id)),

})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App))

App.propTypes = {
  movies: PropTypes.array,
  canPopup: PropTypes.bool,
  user: PropTypes.object,
  setMovies: PropTypes.func,
  setLoginPrompt: PropTypes.func,
  setPopup: PropTypes.func,
}