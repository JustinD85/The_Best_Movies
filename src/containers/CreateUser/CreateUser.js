import React, { Component } from 'react'
import { Redirect, Link } from 'react-router-dom'
import { postData } from '../../api'
import PropTypes from 'prop-types'
import './CreateUser.scss'

class CreateUser extends Component {
  constructor() {
    super()
    this.state = {
      name: '',
      email: '',
      password: '',
      didPost: false,
      error: ''
    }
  }

  handleChange = (e) => {
    const { name, value } = e.target;

    this.setState({
      [name]: value,
    })
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    const { name, email, password } = this.state;
    const user = { name, email, password };

    try {
      await postData('users/new', user)
      this.setState({
        didPost: true
      })
    } catch (error) {
      this.setState({
        error: 'User already exists'
      })
    }
  }
  componentDidMount = () => this.props.setPopup(true)
  render() {
    const { didPost, error } = this.state
    return (
      didPost ?
        <Redirect to='/login' /> :
        <div className='CreateUser'>
          <form onSubmit={this.handleSubmit}>
            <h1 className='create-your-account'>Create Your Account</h1>
            <input
              className='user-input'
              placeholder='Name'
              required type='text'
              onChange={this.handleChange}
              name='name' />
            <input
              className='user-input'
              placeholder='Email'
              required type='email'
              onChange={this.handleChange}
              name='email' />
            <input
              className='user-input'
              placeholder='Password'
              required type='password'
              onChange={this.handleChange}
              name='password' />
            <button
              className='submit-user'
              type='submit'
            >Submit
              </button>
            <Link to='/' onClick={() => this.props.setPopup(false)}>Back</Link>
            {error && <h3>{error}</h3>}
          </form>

        </div>
    )
  }
}

export default CreateUser;

CreateUser.propTypes = {
  setPopup: PropTypes.func,
}