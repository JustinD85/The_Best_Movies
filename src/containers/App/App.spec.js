import React from 'react'
import { shallow } from 'enzyme'
import { App, mapDispatchToProps, mapStateToProps } from './App'
import { Route } from 'react-router-dom'
import { setShouldPromptLogin, setPopup, setMovies } from '../../actions'
import { getMovies, getData } from '../../api/'
import { mockMoviesArray } from '../../mockData/movies'
import CreateUser from '../CreateUser/CreateUser';
jest.mock('../../actions')
jest.mock('../../api/')

describe('App', () => {
  let wrapper
  getMovies.mockReturnValue( mockMoviesArray)
  getData.mockReturnValue({ ok: true })
  window.fetch = jest.fn().mockReturnValue({ ok: true })
  beforeEach(() => {
    wrapper = shallow(<App
      movies={mockMoviesArray}
      canPopup={true}
      setPopup={setPopup}
      setShouldPromptLogin={setShouldPromptLogin}
      setMovies={setMovies}
    />)

  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should invoke setMovies with "now_playing" ', async () => {
    await expect(getMovies).toHaveBeenCalledWith('now_playing')
    await expect(setMovies).toHaveBeenCalled()
  })

  it('should render 4 routes', () => {
    expect(wrapper.find(Route).length).toBe(4)
  })


  describe('mapStateToProps', () => {
    it('should return an array,bool and object', () => {
      const mockState = {
        shouldPromptLogin: false,
        movies: mockMoviesArray,
        user: { name: 'jay' },
        random: 'I am random',
      }

      const expected = {
        canPopup: false,
        movies: mockMoviesArray,
        user: { name: 'jay' },
      }
      const result = mapStateToProps(mockState)
      expect(result).toEqual(expected)

    })
  })

  describe('mapDispatchToProps', () => {
    const dispatchMock = jest.fn()
    const props = mapDispatchToProps(dispatchMock)

    it('should call dispatch set movies is called ', () => {
      const expected = setMovies({ movies: 'yaa' })
      props.setMovies({ movies: 'yaa' })
      expect(dispatchMock).toHaveBeenCalledWith(expected)
    })

    it('should call dispatch setShouldPromptLogin is called ', () => {
      const expected = setShouldPromptLogin(true)
      props.setShouldPromptLogin(true)
      expect(dispatchMock).toHaveBeenCalledWith(expected)
    })

    it('should call dispatch setPopup is called ', () => {
      const expected = setPopup(false)
      props.setPopup(false)
      expect(dispatchMock).toHaveBeenCalledWith(expected)
    })
  })

})