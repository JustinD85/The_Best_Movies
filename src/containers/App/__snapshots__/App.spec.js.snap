// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <Switch>
    //Remove this and make login a popup!!
    <Route
      component={[Function]}
      path="/login"
    />
    <Route
      component={[Function]}
      path="/create-user"
    />
    <Route
      component={[Function]}
      path="/"
    />
  </Switch>
</div>
`;
